CONTRIBUTING.md: add info about updating templates, misc other changes (#38491)
void-linux/void-packages: CONTRIBUTING.md

This document describes how you, as a contributor, can help with adding packages, correcting bugs and adding features to void-packages.

> ## Creating and updating packages in Void by yourself

Reviewer: ```suggestion
## Creating, updating and modifying packages in Void by yourself
```
or something similar that includes other modifications

If you really want to get a new package or package update into Void Linux, we recommend you contribute it yourself.

We provide a [comprehensive Manual](./Manual.md) on how to create new packages.
There's also a [manual for xbps-src](./README.md), which is used to build package files from templates.

For this guide, we assume you have basic knowledge about [git](http://git-scm.org), as well as a [GitHub Account](http://github.com).

To get started, [fork](https://help.github.com/articles/fork-a-repo) the void-linux `void-packages` git repository on GitHub and clone it:

    # with SSH set up for GitHub (recommended)
    $ git clone git@github.com:<user>/void-packages.git
    # without SSH set up for GitHub
    $ git clone https://github.com/<user>/void-packages.git

To keep your forked repository up to date, setup the `upstream` remote to pull in new changes:

    $ git remote add upstream https://github.com/void-linux/void-packages.git
    $ git pull --rebase upstream master

This can also be done with the `github-cli` tool:

    $ gh repo fork void-linux/void-packages
    $ gh repo clone <user>/void-packages

> This automatically sets up the `upstream` remote, so `git pull --rebase upstream master` can still be used to keep your fork up-to-date.

Reviewer: Are you sure the remote is always called upstream after the clone? I think I remember some issues with remote naming, but I don't know what it was.

Author: yes

```
$ gh repo clone --help
...
If the repository is a fork, its parent repository will be added as an additional
git remote called "upstream". The remote name can be configured using `--upstream-remote-name`.
The `--upstream-remote-name` option supports an "@owner" value which will name
the remote after the owner of the parent repository.
...
```

Reviewer: That sounds like it's called upstream only if it's a fork and we are cloning void-linux/void-packages which isn't a fork

    $ git remote add upstream https://github.com/void-linux/void-packages.git
    $ git pull --rebase upstream master

This can also be done with the `github-cli` tool:

>     $ gh repo fork void-linux/void-packages
>     $ gh repo clone <user>/void-packages

Reviewer 2: ```suggestion
    $ gh repo fork --clone void-linux/void-packages
```

Author: I'm aware of that, but I think keeping it as two commands is good because it's clearer what someone who already has forked the repo can do

Reviewer 2: Ah, that makes sense. But, as a side note, the one-liner would also clone the fork even if it's already been created.

Author: does it properly handle having forks in multiple namespaces? e.g. if i had classabbyamp/void-packages and foobar-org/void-packages, will it show a picker or something?

Reviewer 2: I did a quick test and the answer is no. It only takes into account the fork in the main authorized account (classabbyamp/void-packages in your example).

Author: ok, I think I prefer the two-step method then. if someone is experienced with github-cli, maybe they can use the one-step method

    $ xnew pkgname subpkg1 subpkg2 ...

Templates must have the name `void-packages/srcpkgs/<pkgname>/template`, where `pkgname` is the same as the `pkgname` variable in the template.

For deeper insights on the contents of template files, please read the [manual](./Manual.md), and be sure to browse the existing template files in the `srcpkgs` directory of this repository for concrete examples.

### Updating a template

At minimum, a template update will consist of changing `version` and `checksum`, if there was an upstream version change, and/or `revision`, if a template-specific change (e.g. patch, correction, etc.) is needed.
Other changes to the template may be needed depending on what changes the upstream has made.

The checksum can be updated automatically with the `xgensum` helper from the [xtools](https://github.com/leahneukirchen/xtools) package:

    $ xgensum -i <pkgname>

### Committing your changes

After making your changes, please check that the package builds successfully. From the top level directory of your local copy of the `void-packages` repository, run:

    $ ./xbps-src pkg <pkgname>

> Your package must build successfully for at least x86, but we recommend trying to build for armv* as well, e.g.:

Reviewer: ```suggestion
Your package must build successfully for at least x86, but we recommend also trying a cross-build for armv6* e.g.:
```
armv6 often has more issues than armv7 and we should mention this is a cross build to prevent confusion

    $ ./xbps-src -a armv7l pkg <pkgname>

> When building for a native architecture (`x86_64*`, `i686`), building with the `-Q` flag or with `XBPS_CHECK_PKGS=yes` set in `etc/conf` (to run the check phase) is strongly encouraged.

Reviewer: native isn't always x86_64* and i686, could we somehow mention this?

> New packages and updates will not be accepted unless they have been runtime tested.

Reviewer: Because this is right after mentioning the check phase, some people might thing runtime testing is just running check.

When you've finished working on the template file, please check it with `xlint` helper from the [xtools](https://github.com/leahneukirchen/xtools) package:

    $ xlint template

If `xlint` reports any issues, resolve them before committing.

Once you have made and verified your changes to the package template and/or other files, make one commit per package (including all changes to its sub-packages). Each commit message should have one of the following formats:

* for new packages, use `New package: <pkgname>-<version>` ([example](https://github.com/void-linux/void-packages/commit/176d9655429188aac10cd229827f99b72982ab10)).

* for package updates, use `<pkgname>: update to <version>.` ([example](https://github.com/void-linux/void-packages/commit/b6b82dcbd4aeea5fc37a32e4b6a8dd8bd980d5a3)).

* for template modifications without a version change, use `<pkgname>: <reason>` ([example](https://github.com/void-linux/void-packages/commit/8b68d6bf1eb997cd5e7c095acd040e2c5379c91d)).

> * for package removals, use `<pkgname>: remove package` and include the removal reason in the commit body ([example](https://github.com/void-linux/void-packages/commit/3837665503241835bddb73be14d269cf9dbc6020)).

Reviewer: Most of these examples are very old and in some cases, the templates they modify would be considered wrong (lint, different variables, etc.). New people might try to copy things they see in these templates, so we should try to find newer examples.

The checksum can be updated automatically with the `xgensum` helper from the [xtools](https://github.com/chneukirchen/xtools) package:

    $ xgensum -i <pkgname>

### Committing your changes

When you've finished working on the template file, please check it with `xlint` helper from the [xtools](https://github.com/chneukirchen/xtools) package:

    $ xlint template

If `xlint` reports any issues, resolve them before committing.

Once you have made and verified your changes to the package template and/or other files, make one commit per package (including all changes to its sub-packages). Each commit message should have one of the following formats:

* for new packages, use `New package: <pkgname>-<version>` ([example](https://github.com/void-linux/void-packages/commit/176d9655429188aac10cd229827f99b72982ab10)).

* for package updates, use `<pkgname>: update to <version>.` ([example](https://github.com/void-linux/void-packages/commit/b6b82dcbd4aeea5fc37a32e4b6a8dd8bd980d5a3)).

* for template modifications without a version change, use `<pkgname>: <reason>` ([example](https://github.com/void-linux/void-packages/commit/8b68d6bf1eb997cd5e7c095acd040e2c5379c91d)).

> * for package removals, use `<pkgname>: remove package` ([example](https://github.com/void-linux/void-packages/commit/83784632d94deee5d038c8e1c4c1dffa922fca21)).

Reviewer: and ideally add the reason to the commit body

The checksum can be updated automatically with the `xgensum` helper from the [xtools](https://github.com/chneukirchen/xtools) package:

    $ xgensum -i <pkgname>

### Committing your changes

When you've finished working on the template file, please check it with `xlint` helper from the [xtools](https://github.com/chneukirchen/xtools) package:

    $ xlint template

If `xlint` reports any issues, resolve them before committing.

Once you have made and verified your changes to the package template and/or other files, make one commit per package (including all changes to its sub-packages). Each commit message should have one of the following formats:

* for new packages, use `New package: <pkgname>-<version>` ([example](https://github.com/void-linux/void-packages/commit/176d9655429188aac10cd229827f99b72982ab10)).

* for package updates, use `<pkgname>: update to <version>.` ([example](https://github.com/void-linux/void-packages/commit/b6b82dcbd4aeea5fc37a32e4b6a8dd8bd980d5a3)).

* for template modifications without a version change, use `<pkgname>: <reason>` ([example](https://github.com/void-linux/void-packages/commit/8b68d6bf1eb997cd5e7c095acd040e2c5379c91d)).

> * for package removals, use `<pkgname>: remove package` and include the removal reason in the commit body ([example](https://github.com/void-linux/void-packages/commit/83784632d94deee5d038c8e1c4c1dffa922fca21)).

Reviewer 2: Example commit could include a removal reason for reference, e.g. https://github.com/void-linux/void-packages/commit/3837665503241835bddb73be14d269cf9dbc6020

* for `common/shlibs` modifications, use `common/shlibs: <pkgname>` ([example](https://github.com/void-linux/void-packages/commit/613651c91811cb4fd2e1a6be701c87072d759a9f)).

* for changes to any other file, use `<filename>: <reason>` ([example](https://github.com/void-linux/void-packages/commit/e00bea014c36a70d60acfa1758514b0c7cb0627d),
  [example](https://github.com/void-linux/void-packages/commit/93bf159ce10d8e474da5296e5bc98350d00c6c82), [example](https://github.com/void-linux/void-packages/commit/dc62938c67b66a7ff295eab541dc37b92fb9fb78))

> If you want to describe your changes in more detail, add an empty line followed by those details ([example](https://github.com/void-linux/void-packages/commit/f1c45a502086ba1952f23ace9084a870ce437bc6)).

Reviewer 3: > , explain in body separated with empty line from headline

so it explains what aforementioned body is

    $ ./xbps-src -a armv7l pkg <pkgname>

Runtime testing of packages and building with the `-Q` flag or with `XBPS_CHECK_PKGS=yes` set in `etc/conf` is strongly encouraged.

> New packages and updates will not be accepted unless they have been runtime tested.

Reviewer: one line says runtime testing is strongly encouraged, the other one that it's required

Alternatively, if there are issues with your git history, you can make another branch and push it to the existing PR:

    $ git checkout master -b <attempt2>

>     # do changes anew

Reviewer 3: building as root is unsupported ;)
